rules/trackhub.smk

```diff
@@ -183,13 +185,187 @@ rule twobit:
+    
+    source: http://genomewiki.ucsc.edu/index.php/Assembly_Hubs#Cytoband_Track
+    """
+    input:
+        genome = expand("{genome_dir}/{{assembly}}/{{assembly}}.fa", **config),
+        sizes = expand("{genome_dir}/{{assembly}}/{{assembly}}.fa.sizes", **config),
+    output:
+        cytoband_bb = expand("{genome_dir}/{{assembly}}/cytoBandIdeo.bb", **config),
+        cytoband_bd = temp(expand("{genome_dir}/{{assembly}}/cytoBandIdeo.bed", **config)),
+    log:
+        expand("{log_dir}/trackhub/{{assembly}}.cytoband.log", **config)
+    benchmark:
+        expand("{log_dir}/trackhub/{{assembly}}.cytoband.benchmark.txt", **config)[0]
+    conda:
+        "../envs/ucsc.yaml"
+    shell:
+        """
+        cat {input.sizes} | bedSort /dev/stdin /dev/stdout | awk '{{print $1,0,$2,$1,"gneg"}}' > {output.cytoband_bd}
+
+        bedToBigBed -type=bed4 {output.cytoband_bd} -as=../../schemas/cytoBand.as {input.sizes} {output.cytoband_bb} >> {log} 2>&1
+        """
+
+
+def get_masked_regions(contig):
+    masked_regions=''
+
+    masked_seq = str(contig.seq)
+    inMasked = False
+    mmEnd = 0
+
+    length = len(masked_seq)-1
+    for x in range(0, length+1):
+        # mark the starting position of a softmasked region
+        if masked_seq[x].islower() and inMasked == False:
+            mmStart = x + 1
+            inMasked = True
+
+        # mark end position of softmasked region (can be end of contig)
+        elif (not(masked_seq[x].islower()) or x == length) and inMasked == True:
+            mmEnd = x
+            inMasked = False
+
+            # store softmasked region in a bed3 (chr, start, end) file
+            masked_regions += contig.id + '\t' + str(mmStart) + '\t' + str(mmEnd) + '\n'
+
+    return masked_regions
+
+
+rule softmask_track_1:
+    """
+    Generate a track of all softmasked regions
+    
+    source: https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py
+    """
+    input:
+        genome = expand("{genome_dir}/{{assembly}}/{{assembly}}.fa", **config),
+    output:
+        mask_unsorted = temp(expand("{genome_dir}/{{assembly}}/{{assembly}}_softmasking_unsorted.bed", **config)),
+    log:
+        expand("{log_dir}/trackhub/{{assembly}}.softmask1.log", **config)
+    benchmark:
+        expand("{benchmark_dir}/trackhub/{{assembly}}.softmask1.benchmark.txt", **config)[0]
+    threads: 4
+    resources:
+        mem_gb=2
+    run:
+        with open(str(input.genome), "r") as genome_handle, open(str(output.mask_unsorted), "w+") as bed_handle:
+            p = Pool(threads)
+            # seqIO.parse returns contig data.
+            # Each contig is scanned by get_masked_regions (in parallel by imap_unordered).
+            # As soon as a contig is scanned, the output is yielded and written to file.
+            for softmasked_regions_per_contig in p.imap_unordered(get_masked_regions, SeqIO.parse(genome_handle, "fasta")):
+                bed_handle.write(softmasked_regions_per_contig)
```

Comment:
Lol is this really the best way? That's how they do it in MakeHub?

Reply:
yes, but without multiprocessing it. And I think they miss any masked regions at the end of the contig.